barrier.go

```diff
@@ -0,0 +1,273 @@
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		n.resetTimer()
+		return m, nil
+	}
+	return nil, nil
+}
+func (n *idleBarrier) EndBatch(m edge.EndBatchMessage) (edge.Message, error) {
+	return m, nil
+}
+func (n *idleBarrier) Barrier(m edge.BarrierMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		n.resetTimer()
+		return m, nil
+	}
+	return nil, nil
+}
+func (n *idleBarrier) DeleteGroup(m edge.DeleteGroupMessage) (edge.Message, error) {
+	if m.GroupID() == n.group.ID {
+		n.Stop()
+	}
+	return m, nil
+}
+
+func (n *idleBarrier) Point(m edge.PointMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		n.resetTimer()
+		return m, nil
+	}
+	return nil, nil
+}
+
+func (n *idleBarrier) resetTimer() {
+	if n.idle > 0 {
+		n.timer.Reset(n.idle)
+	}
+}
+
+func (n *idleBarrier) emitBarrier() error {
+	nowT := time.Now()
+	n.lastT.Store(nowT)
+	return edge.Forward(n.outs, edge.NewBarrierMessage(n.group, nowT))
+}
+
+func (n *idleBarrier) idleHandler() {
+	defer n.wg.Done()
+	for {
+		select {
+		case <-n.timer.C:
+			n.emitBarrier()
+			n.resetTimer()
+		case <-n.stopC:
+			return
+		}
+	}
+}
+
+type periodicBarrier struct {
+	name  string
+	group edge.GroupInfo
+
+	lastT  atomic.Value
+	ticker *time.Ticker
+	wg     sync.WaitGroup
+	outs   []edge.StatsEdge
+	stopC  chan bool
+}
+
+func newPeriodicBarrier(name string, group edge.GroupInfo, period time.Duration, outs []edge.StatsEdge) *periodicBarrier {
+	r := &periodicBarrier{
+		name:   name,
+		group:  group,
+		lastT:  atomic.Value{},
+		ticker: time.NewTicker(period),
+		wg:     sync.WaitGroup{},
+		outs:   outs,
+		stopC:  make(chan bool, 1),
+	}
+
+	r.Init()
+
+	return r
+}
+
+func (n *periodicBarrier) Init() {
+	n.lastT.Store(time.Time{})
+	n.wg.Add(1)
+
+	go n.periodicEmitter()
+}
+
+func (n *periodicBarrier) Stop() {
+	n.stopC <- true
+	n.ticker.Stop()
+	n.wg.Wait()
+}
+
+func (n *periodicBarrier) BeginBatch(m edge.BeginBatchMessage) (edge.Message, error) {
+	return m, nil
+}
+func (n *periodicBarrier) BatchPoint(m edge.BatchPointMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		return m, nil
+	}
+	return nil, nil
+}
+func (n *periodicBarrier) EndBatch(m edge.EndBatchMessage) (edge.Message, error) {
+	return m, nil
+}
+func (n *periodicBarrier) Barrier(m edge.BarrierMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		return m, nil
+	}
+	return nil, nil
+}
+func (n *periodicBarrier) DeleteGroup(m edge.DeleteGroupMessage) (edge.Message, error) {
+	if m.GroupID() == n.group.ID {
+		n.Stop()
+	}
+	return m, nil
+}
+
+func (n *periodicBarrier) Point(m edge.PointMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		return m, nil
+	}
+	return nil, nil
+}
+
+func (n *periodicBarrier) emitBarrier() error {
+	nowT := time.Now()
+	n.lastT.Store(nowT)
+	return edge.Forward(n.outs, edge.NewBarrierMessage(n.group, nowT))
```

Comment:
Instead of using an atomic.Value on `n.lastT` I think using a sync.Mutex will be helpful, since you will want to hold that lock when your are writing to the `n.outs` since you do not want to concurrently write the Barrier message with any other message.

Reply:
The channels prevent a race condition during write but you still need to guarantee that you do not mix messages in a invalid order. For example during a batch write only batchpoint messages should exist between begin and end messages.